pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,330 @@
+	// TODO: get this from the image metadata
+	scriptsDestDir := filepath.Join(getDestination(config), "scripts")
+	sourcesDestDir := filepath.Join(getDestination(config), "src")
+
+	buffer.WriteString(fmt.Sprintf("FROM %s\n", builder.config.BuilderImage))
+	buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+
+	generatedLabels := util.GenerateOutputImageLabels(builder.sourceInfo, builder.config)
+	if len(generatedLabels) > 0 {
+		buffer.WriteString("LABEL")
+		for k, v := range generatedLabels {
+			buffer.WriteString(fmt.Sprintf(" %s=%s", k, v))
+		}
+		buffer.WriteString("\n")
+	}
+
+	// only COPY scripts dir if required scripts are present, i.e. the dir is not empty;
+	// even if the "scripts" dir exists, the COPY would fail if it was empty
+	scriptsProvided, fileNames := checkValidDirWithContents(filepath.Join(builder.config.WorkingDir, builder.uploadScriptsDir))
+	//glog.V(2).Infof("found %v files in script dir %s", fileNames, builder.uploadScriptsDir)
+	assembleProvided := false
+	runProvided := false
+	for _, f := range fileNames {
+		glog.V(2).Infof("found script file %s", f.Name())
+		if f.Name() == "run" {
+			runProvided = true
+		} else if f.Name() == "assemble" {
+			assembleProvided = true
+		}
+		if runProvided && assembleProvided {
+			break
+		}
+	}
+
+	if scriptsProvided {
+		glog.V(2).Infof("The scripts are included in %q directory", builder.uploadScriptsDir)
+		buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(builder.uploadScriptsDir), sanitize(filepath.ToSlash(scriptsDestDir))))
+	} else {
+		// if an err on reading or opening dir, can't copy it
+		glog.V(2).Infof("Could not gather scripts from the directory %q", builder.uploadScriptsDir)
+	}
+
+	env := sti.CreateBuildEnvironment(builder.config.WorkingDir, config.Environment)
+	if len(env) > 0 {
+		buffer.WriteString("ENV")
+		for _, v := range env {
+			buffer.WriteString(fmt.Sprintf(" %s", sanitize(v)))
+		}
+		buffer.WriteString("\n")
+	}
+
+	buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(builder.uploadSrcDir), sanitize(filepath.ToSlash(sourcesDestDir))))
+
+	glog.V(2).Infof("Processing secret inputs: %#v", config.Injections)
+	config.Injections = util.FixInjectionsWithRelativePath(config.ImageWorkDir, config.Injections)
+	glog.V(2).Infof("Processed secret inputs: %#v", config.Injections)
+
+	for _, injection := range config.Injections {
+		//src := strings.TrimPrefix(injection.Source, builder.config.WorkingDir)
+		src := filepath.Join(api.Injections, injection.Source)
+		buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(src), sanitize(injection.Destination)))
+	}
+
+	if assembleProvided {
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(scriptsDestDir, "assemble"))))
+	} else {
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(inImageScriptsDir, "assemble"))))
```

Comment:
So as part of "the challenge is that since we can't inspect the builder image", we have to just  assume the assemble script is in the builder image .... and if it is not, the failure will be such that the user can sort that out

Could it be beneficial to add a comment to the Dockerfile explaining we've made an assumption the file is there, you need to make sure it is there, etc.

Reply:
sure, will add.